pypeit/spectrographs/keck_nirspec.py

```diff
@@ -6,24 +7,1124 @@
+            Detector metadata parameters for one or more detectors.
+        raw_img : `numpy.ndarray`_
+            Raw image for this detector.  Shape is 2D if a single detector image
+            is read and 3D if multiple detectors are read.  E.g., the image from
+            the first detector in the tuple is accessed using ``raw_img[0]``.
+        hdu : `astropy.io.fits.HDUList`_
+            Opened fits file
+        exptime : :obj:`float`
+            Exposure time *in seconds*.
+        rawdatasec_img : `numpy.ndarray`_
+            Data (Science) section of the detector as provided by setting the
+            (1-indexed) number of the amplifier used to read each detector
+            pixel. Pixels unassociated with any amplifier are set to 0.  Shape
+            is identical to ``raw_img``.
+        oscansec_img : `numpy.ndarray`_
+            Overscan section of the detector as provided by setting the
+            (1-indexed) number of the amplifier used to read each detector
+            pixel. Pixels unassociated with any amplifier are set to 0.  Shape
+            is identical to ``raw_img``.
+        """
+        # Check extension and then open
+        self._check_extensions(raw_file)
+        hdu = io.fits_open(raw_file)
+
+        # Validate the entered (list of) detector(s)
+        nimg, _det = self.validate_det(det)
+
+        # Grab the detector or mosaic parameters
+        mosaic = None if nimg == 1 else self.get_mosaic_par(det, hdu=hdu)
+        detectors = [self.get_detector_par(det, hdu=hdu)] if nimg == 1 else mosaic.detectors
+
+        # Grab metadata from the header
+        # NOTE: These metadata must be *identical* for all images when reading a
+        # mosaic
+        headarr = self.get_headarr(hdu)
+
+        # save filter being used:
+        self.filter1 = self.get_meta_value(headarr, 'filter1')
+
+        # Exposure time (used by RawImage)
+        # NOTE: This *must* be (converted to) seconds.
+        exptime = self.get_meta_value(headarr, 'exptime')
+
+        # Rawdatasec, oscansec images
+        binning = self.get_meta_value(headarr, 'binning')
+        # NOTE: This means that `specaxis` must be the same for all detectors in
+        # a mosaic
+        if detectors[0]['specaxis'] == 1:
+            binning_raw = (',').join(binning.split(',')[::-1])
+        else:
+            binning_raw = binning
+
+        raw_img = [None]*nimg
+        rawdatasec_img = [None]*nimg
+        oscansec_img = [None]*nimg
+        for i in range(nimg):
+
+            # Raw image
+            raw_img[i] = hdu[detectors[i]['dataext']].data.astype(float)
+            # Raw data from some spectrograph (i.e. FLAMINGOS2) have an addition
+            # extention, so I add the following two lines. It's easier to change
+            # here than writing another get_rawimage function in the
+            # spectrograph file.
+            # TODO: This feels dangerous, but not sure what to do about it...
```

Comment:
Can we simply throw an error if the ndim isn't 2 to begin with? Are there really cases where the number of dimensions is not 2?

Reply:
Yes, that's probably the safer move because who knows what can happen down the line. I've made the correction